include/oneapi/dpl/pstl/tuple_impl.h

```diff
@@ -487,38 +456,100 @@ struct tuple<>
     {
         return *this;
     }
-    friend constexpr bool
-    operator==(const tuple& __lhs, const tuple& __rhs)
-    {
-        return tuple_type{} == tuple_type{};
-    }
-    friend constexpr bool
-    operator!=(const tuple& __lhs, const tuple& __rhs)
-    {
-        return tuple_type{} != tuple_type{};
-    }
-    friend constexpr bool
-    operator<(const tuple& __lhs, const tuple& __rhs)
-    {
-        return tuple_type{} < tuple_type{};
-    }
-    friend constexpr bool
-    operator<=(const tuple& __lhs, const tuple& __rhs)
-    {
-        return tuple_type{} <= tuple_type{};
-    }
-    friend constexpr bool
-    operator>(const tuple& __lhs, const tuple& __rhs)
-    {
-        return tuple_type{} > tuple_type{};
-    }
-    friend constexpr bool
-    operator>=(const tuple& __lhs, const tuple& __rhs)
-    {
-        return tuple_type{} >= tuple_type{};
-    }
 };
 
+template <typename T>
+struct __get_std_tuple_t_impl
+{
+    using type = void;
+};
+
+template <typename... T>
+struct __get_std_tuple_t_impl<std::tuple<T...>>
+{
+    using type = std::tuple<T...>;
+};
+
+template <typename... T>
+struct __get_std_tuple_t_impl<oneapi::dpl::__internal::tuple<T...>>
+{
+    using type = std::tuple<T...>;
+};
+
+template <typename T>
+using __get_std_tuple_t = typename __get_std_tuple_t_impl<std::decay_t<T>>::type;
+
+template <typename T>
+using __is_tuple_type = std::negation<std::is_void<__get_std_tuple_t<T>>>;
+
+template <typename T>
+using __is_std_tuple_type = std::is_same<__get_std_tuple_t<T>, std::decay_t<T>>;
+
+// Require that both are tuple types, but at least one type is a oneDPL internal tuple, and that the number of tuple
+// elements are the same
+template <typename Tuple1, typename Tuple2>
+using __comparable_tuple_types = std::conjunction<
+    __is_tuple_type<Tuple1>, __is_tuple_type<Tuple2>,
+    std::negation<std::conjunction<__is_std_tuple_type<Tuple1>, __is_std_tuple_type<Tuple2>>>,
+    std::bool_constant<std::tuple_size<std::decay_t<Tuple1>>::value == std::tuple_size<std::decay_t<Tuple2>>::value>>;
+
+// These binary comparison overloads required because the template arguments require exact matches to participate in
+// overload resolution. Implicit conversion to std::tuple does not provide these for free. The static cast should be
+// without performance penalty if the casting to its own type.
+template <typename Tuple1, typename Tuple2,
+          typename = std::enable_if_t<__comparable_tuple_types<Tuple1, Tuple2>::value>>
+constexpr bool
+operator==(Tuple1&& __lhs, Tuple2&& __rhs)
+{
+    return static_cast<__get_std_tuple_t<Tuple1>>(std::forward<Tuple1>(__lhs)) ==
+           static_cast<__get_std_tuple_t<Tuple2>>(std::forward<Tuple2>(__rhs));
```

Comment:
I double checked on [Godbolt](https://godbolt.org/z/ccc3n7q8s) and this does not introduce a copy for the cast of a `std::tuple` to a `std::tuple`. I don't have anything to add; I just wanted to verify that there wasn't an unnecessary copy being introduced.

Reply:
I'm still working it out.  I think it may inject an extra copy for lvalue refs... still a work in progress.